lua/entities/gmod_wire_expression2/init.lua

```diff
@@ -584,3 +584,51 @@ function MakeWireExpression2(player, Pos, Ang, model, buffer, name, inputs, outp
 	return self
 end
 duplicator.RegisterEntityClass("gmod_wire_expression2", MakeWireExpression2, "Pos", "Ang", "Model", "_original", "_name", "_inputs", "_outputs", "_vars", "inc_files", "filepath")
+
+--------------------------------------------------
+-- Emergency shutdown (beta testing so far)
+--------------------------------------------------
+local average_ram = 0
+local enable = CreateConVar( 
+	"wire_expression2_ram_emergency_shutdown_enable", "0", {FCVAR_ARCHIVE}, 
+	"Enable/disable the emergency shutdown feature. Requires map reload after change." )
+
+local average_halt_multiplier = CreateConVar( 
+	"wire_expression2_ram_emergency_shutdown_spike", "4", {FCVAR_ARCHIVE}, 
+	"if (current_ram > average_ram * spike_convar) then shut down all E2s" )
+
+local halt_max_amount = CreateConVar( 
+	"wire_expression2_ram_emergency_shutdown_total", "512", {FCVAR_ARCHIVE}, 
+	"This is in kilobytes, if (current_ram > total_convar) then shut down all E2s" )
+
+if enable:GetBool() then
+	hook.Remove( "Think", "wire_expression2_emergency_shutdown" ) -- remove old hook
```

Comment:
https://github.com/garrynewman/garrysmod/blob/974f6ce8ed8a50f854179507435e3d4fe1176f46/garrysmod/lua/includes/modules/hook.lua#L32 rather - notice that it will overwrite any previous entry with the same event and name.